vach/src/writer/builder/config.rs

```diff
@@ -3,34 +3,27 @@ use std::io;
 use ed25519_dalek as esdalek;
 use std::fmt::Debug;
 
-type OptionalCallback = Option<Box<dyn Fn(&str, usize, &RegistryEntry)>>;
-
 /// Allows for the customization of valid `vach` archives during their construction.
 /// Such as custom `MAGIC`, custom `Header` flags and signing by providing a keypair.
-pub struct BuilderConfig {
+pub struct BuilderConfig<'a> {
 	/// Used to write a unique magic sequence into the write target.
 	pub magic: [u8; crate::MAGIC_LENGTH],
 	/// Flags to be written into the `Header` section of the write target.
 	pub flags: Flags,
 	/// An optional keypair. If a key is provided, then the write target will have signatures for tamper verification.
 	pub keypair: Option<esdalek::Keypair>,
-	/// An optional callback that is called every time a `Leaf` finishes processing.
-	/// The type signature is:
-	/// ```ignore
-	/// type OptionalCallback = Option<Box<dyn Fn(id: &str, glob_size: usize, reg_entry: &RegistryEntry)>>;
-	/// ```
+	/// An optional callback that is called every time a [Leaf](crate::builder::Leaf) finishes processing.
+	/// The callback get passed to it: the leaf's id, the number of bytes written and the generated registry entry. Respectively.
+	/// > **To avoid** the `implementation of "FnOnce" is not general enough` error consider adding types to the closure's parameters, as this is a type inference error. Rust somehow cannot infer enough information, [link](https://www.reddit.com/r/rust/comments/ntqu68/implementation_of_fnonce_is_not_general_enough/).
 	///
 	/// Usage:
 	/// ```ignore
 	/// let builder_config = BuilderConfig::new()
 	/// ```
-	pub progress_callback: OptionalCallback,
-	/// Reserve some space in the registry section of an archive, allowing entries to be added later in the archives lifetime.
-	/// Thus making the archive mutable. Passing `None` or `Some(x) (where x < RegistryEntry::SIZE)`, results in an immutable archive
-	pub reserved_reg_space: bool
+	pub progress_callback: Option<&'a dyn Fn(&str, usize, &RegistryEntry)>
```

Comment:
Maybe we should box this function so we can get rid of the lifetime

Reply:
The benefit here is that one can define a closure and use it in multiple `BuilderConfig`s this is because it now takes a reference rather than ownership. It adds some flexibility, plus Rust lifetimes are very easy to use, just specify `'static`. The lifetime just adds some extra flexibility to those who wants to build structs on the `BuilderConfig`